### PR TITLE
[luci] Add pass to replace mul/add with depthwise conv

### PR DESCRIPTION
**Pull request by jinevening**
This adds a pass to replace mul/add with depthwise conv

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Draft PR: #5244

### CODE REVIEW

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  assert(gamma->rank() == 1);
+  auto channel_size = gamma->dim(0).value();
+
+  auto weights = gamma->graph()->nodes()->create<luci::CircleConst>();
+  weights->dtype(loco::DataType::FLOAT32);
+  weights->rank(4);
+  weights->dim(0).set(1);
+  weights->dim(1).set(1);
+  weights->dim(2).set(1);
+  weights->dim(3).set(channel_size);
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:35]: Could you please add some comments about this conversion?

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+luci::CircleConst *create_bias_from_beta(luci::CircleConst *beta)
+{
+  assert(beta->rank() == 1);
+  auto channel_size = beta->dim(0).value();
+
+  auto bias = beta->graph()->nodes()->create<luci::CircleConst>();
+  bias->dtype(loco::DataType::FLOAT32);
+  bias->rank(1);
+  bias->dim(0).set(channel_size);
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:54]: Here too

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  luci::CircleConst *beta = nullptr;
+  luci::CircleConst *gamma = nullptr;
+
+  if (!is_batchnorm_add(add, mul, beta))
+    return false;
+
+  if (loco::succs(mul).size() != 1)
+    return false;
+
+  if (!is_batchnorm_mul(mul, pred_node, gamma))
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:168]: maybe check `pred_node` shape and dtype(?)... `create_weights_from_gamma` creates fixed rank 4 CircleConst without any shape check...

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    return false;
+
+  if (beta->dtype() != loco::DataType::FLOAT32 || gamma->dtype() != loco::DataType::FLOAT32)
+    throw std::runtime_error("ReplaceMulAddWithDepthwiseConvPass only supports Float32 model");
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:172]: I think `!FLOAT32` can happen so maybe just return false?

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp (hunk 1 above)
Comment (jinevening) [reply compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:36]: Added comments in both weights and bias.

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp (hunk 3 above)
Comment (jinevening) [reply compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:170]: I missed it. Thanks.

Added check for pred_node.

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp (hunk 4 above)
Comment (jinevening) [reply compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp]: Changed to return false. Thanks.

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Comment by seanshpark** [comment]
Please resolve conflicts

**Comment by jinevening** [comment]
> Please resolve conflicts

Resolved. PTAL :)

**Review by llFreetimell — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM

**Pull request merged**